examples/getting_started/tools_and_agents.ipynb

```diff
@@ -0,0 +1,583 @@
+    "            \"tools\": [self.web_search, extract_content],\n",
+    "            \"computed_fields\": {\n",
+    "                \"current_date\": datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
+    "            },\n",
+    "        }\n",
+    "\n",
+    "    async def _step(self, question: str):\n",
+    "        response = await self._stream(question)\n",
+    "        tools_and_outputs = []\n",
+    "        async for chunk, tool in response:\n",
+    "            if tool:\n",
+    "                print(f\"using {tool._name()} tool with args: {tool.args}\")\n",
+    "                tools_and_outputs.append((tool, tool.call()))\n",
+    "            else:\n",
+    "                print(chunk.content, end=\"\", flush=True)\n",
+    "        if response.user_message_param:\n",
+    "            self.messages.append(response.user_message_param)\n",
+    "        self.messages.append(response.message_param)\n",
+    "        if tools_and_outputs:\n",
+    "            self.messages += response.tool_message_params(tools_and_outputs)\n",
+    "            await self._step(\"\")\n",
+    "\n",
+    "    async def run(self):\n",
+    "        while True:\n",
+    "            question = input(\"(User): \")\n",
+    "            if question == \"exit\":\n",
+    "                break\n",
+    "            print(\"(Assistant): \", end=\"\", flush=True)\n",
+    "            await self._step(question)\n",
+    "            print()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This implementation includes:\n",
+    "\n",
+    "1. The `_stream` method, which sets up the LLM call with the necessary tools and computed fields.\n",
+    "2. The `_step` method, which processes the LLM response, handles tool calls, and updates the conversation history.\n",
+    "3. The `run` method, which implements the main interaction loop for the agent.\n",
+    "\n",
+    "The `@openai.call` and `@prompt_template` decorators are used to set up the LLM interaction and define the prompt for the agent.\n",
+    "\n",
+    "Note how we're passing `self.web_search` as a tool, which allows the LLM to use our custom web search method that has access to the agent's state."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 7. Running and Testing the Agent\n",
+    "\n",
+    "Now that we have implemented our WebSearchAgent, let's run and test it:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-09-11T08:23:30.442635Z",
+     "start_time": "2024-09-11T08:22:56.224422Z"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "(Assistant): using web_search tool with args: {'queries': ['large language models overview', 'applications of large language models', 'future of large language models', 'impact of large language models on AI']}\n",
```

Comment:
nit: something weird happening here where the input isn't being saved as output, so maybe include the input in the cell as a comment so it's more clear what you asked?